src/sieve.py

```diff
@@ -16,5 +16,18 @@ def sieve(n: int) -> list[int]:
     primes = []
 
     # FIXME: fill out this bit
-
+    while candidates:
+        i=1
+        len_can=len(candidates)
+        while i < len_can:
+            p=candidates[0]
+            if candidates[i]%p==0:
+                candidates.remove(candidates[i])
+                len_can-=1
+            else:
+                i+=1
+        primes.append(candidates[0])
+        candidates.remove(candidates[0])
```

Comment:
Keeping track of the length of candidates like this is a little fragile. You have the same information twice, in `len_ca` and in `len(candidates)`, and bad things happen if they are out of sync. It is better to keep information in a single place so such dangers no longer exist. In your code, you could just use `len(candidates)` in the test in the inner while loop.

I wouldn't recommend using remove on lists if you can avoid it. It is a slow operation (and later in the class I will explain why). It is faster to create a whole new list of candidates than it is to remove candidates one at a time.

You could instead do something like this:

```python
def sieve(n: int) -> list[int]:
    primes = []
    candidates = list(range(2, n+1))
    while candidates:
        p, new_candidates = candidates[0], []
        primes.append(p)
        for cand in candidates:
            if cand % p != 0:
                new_candidates.append(cand)
        candidates = new_candidates
    return primes
```

or with a list-comprehension:

```python
def sieve(n: int) -> list[int]:
    primes = []
    candidates = list(range(2, n+1))
    while candidates:
        primes.append(candidates[0])
        candidates = [cand for cand in candidates if cand % primes[-1] != 0]
    return primes
```

It is doing exactly the same as your solution, but just building a new candidates list with the surviving members, rather than deleting the excluded members one at a time.